evm-validate/src/lib.rs

```diff
@@ -417,3 +468,764 @@ fn validate_transaction_hashes(
+        let log = Log::new(address, topics, log_data).expect("log is invalid");
+        tx_logs.push(log);
+    };
+    // store the last tx logs in the mapping
+    logs_by_block_num_and_tx_idx.insert((current_block_num, current_tx_idx), tx_logs); 
+
+    // CREATE A TRANSACTION MAPPING
+
+    let tx_array = extract_transaction_cols_as_arrays(transactions)?;  
+    // get first tx block num
+    let mut current_block_num = tx_array.block_number.value(0);
+    // initialize a map to store transaction root by block num
+    let mut transactions_root_by_block_num_mapping = BTreeMap::<u64, FixedBytes<32>>::new();
+    // initialize a map to store receipts by block num
+    let mut receipts_root_by_block_num_mapping = BTreeMap::<u64, FixedBytes<32>>::new();
+    // initialize a vec to store tx envelopes for a tx
+    let mut block_tx_envelopes = Vec::<TxEnvelope>::with_capacity(200);
+    // initialize a vec to store receipts for a tx
+    let mut block_tx_receipts = Vec::<ReceiptEnvelope>::with_capacity(200);
+    // initialize an empty vec of logs, used if the tx failed or doesn't have logs
+    let empty_logs = Vec::<Log>::new();
+
+    let tx_iterators = tx_array
+        .block_number.iter()
+        .zip(tx_array.gas_limit.iter())
+        .zip(tx_array.gas_price.iter())
+        .zip(tx_array.hash.iter())
+        .zip(tx_array.input.iter())
+        .zip(tx_array.nonce.iter())
+        .zip(tx_array.to.iter())
+        .zip(tx_array.tx_index.iter())
+        .zip(tx_array.value.iter())
+        .zip(tx_array.v.iter())
+        .zip(tx_array.r.iter())
+        .zip(tx_array.s.iter())
+        .zip(tx_array.max_priority_fee_per_gas.iter())
+        .zip(tx_array.max_fee_per_gas.iter())
+        .zip(tx_array.chain_id.iter())
+        .zip(tx_array.cumulative_gas_used.iter())
+        .zip(tx_array.contract_address.iter())
+        .zip(tx_array.logs_bloom.iter())
+        .zip(tx_array.tx_type.iter())
+        .zip(tx_array.status.iter())
+        .zip(tx_array.sighash.iter())
+        .zip(tx_array.access_list.iter())
+        .zip(tx_array.max_fee_per_blob_gas.iter())
+        .zip(tx_array.blob_versioned_hashes.iter());
+    
+    // iterate over transactions rows
+    for (((((((((((((((((((((((
+        tx_block_nums_opt
+        , tx_gas_limit_opt)
+        , tx_gas_price_opt)
+        , tx_hash_opt)
+        , tx_input_opt)
+        , tx_nonce_opt)
+        , tx_to_opt)
+        , tx_tx_idx_opt)
+        , tx_value_opt)
+        , tx_v_opt)
+        , tx_r_opt)
+        , tx_s_opt)
+        , tx_max_priority_fee_per_gas_opt)
+        , tx_max_fee_per_gas_opt)
+        , tx_chain_id_opt)
+        , tx_cumulative_gas_used_opt)
+        , tx_contract_address_opt)
+        , tx_logs_bloom_opt)
+        , tx_type_opt) 
+        , tx_status_opt)
+        , tx_sighash_opt)
+        , tx_access_list_opt)
+        , tx_max_fee_per_blob_gas_opt)
+        , tx_blob_versioned_hashes_opt) in tx_iterators {
```

Comment:
Even though this is messy, the iterators are more performant than if I make a for loop in the length of the array. For instance, the index lookup must check if the index is not out-of-bound of the array..

Reply:
don't think it will make much of difference in practice

Reply:
I've tried using a for loop to test. When you use iter, it returns an Option if the value is empty, which is good for many fields. I already went and implemented this approach; I don't want to change it.